lib/Interface/Container.cpp

```diff
@@ -0,0 +1,287 @@
+    : offset(other.offset), size(other.size) {
+  for (size_t i = 0; i < N; i++) {
+    this->sizes[i] = other.sizes[i];
+  }
+  setStrides();
+  allocated = new T[size];
+  aligned = allocated;
+  for (size_t i = 0; i < size; i++) {
+    aligned[i] = other.aligned[i];
+  }
+}
+
+/**
+ * @brief Copy Assignment Operator..
+ * - Check if they are the same object.
+ * - Copy `offset` and `size` directly.
+ * - Elementwise copy `sizes`.
+ * - Calculate the `strides`.
+ * - Free the data space of this object to avoid memory leaks.
+ * - Allocate new space and deep copy.
+ * @tparam T represents the datatype to be used
+ * @tparam N represents the number of dimensions
+ */
+template <typename T, std::size_t N>
+MemRef<T, N> &MemRef<T, N>::operator=(const MemRef<T, N> &other) {
+  if (this != &other) {
+    this->offset = other.offset;
+    this->size = other.size;
+    for (size_t i = 0; i < N; i++) {
+      this->sizes[i] = other.sizes[i];
+    }
+    setStrides();
+    // Free the original aligned and allocated space.
+    delete[] allocated;
+    // Allocate new space and deep copy.
+    T *ptr = new T[size];
+    for (size_t i = 0; i < size; i++) {
+      ptr[i] = other.aligned[i];
+    }
+    aligned = ptr;
+    allocated = ptr;
+  }
+  return *this;
+}
+
+/**
+ * @brief Move Constructor. This constructor is used to initialize a MemRef object from a rvalue.
+ * The move constructor steals the resources of the original object.
+ * Note that the original object no longer owns the members and spaces.
+ * - Steal members from the original object.
+ * - Assign the NULL pointer to the original aligned and allocated members to
+ *   avoid the double free error.
+ * @tparam T represents the datatype to be used
+ * @tparam N represents the number of dimensions
+ */
+template <typename T, std::size_t N>
+MemRef<T, N>::MemRef(MemRef<T, N> &&other) noexcept
+    : allocated(other.allocated), aligned(other.aligned), offset(other.offset),
+      size(other.size) {
+  std::swap(this->sizes, other.sizes);
+  std::swap(this->strides, other.strides);
+  // Assign the NULL pointer to the original aligned and allocated members to
+  // avoid the double free error.
+  other.allocated = other.aligned = nullptr;
+}
+
+/**
+ * @brief Move Assignment Operator. Note that the original object no longer owns the members and spaces.
+ * - Check if they are the same object.
+ * - Free the data space of this object to avoid memory leaks.
+ * - Steal members from the original object.
+ * - Assign the NULL pointer to the original aligned and allocated members to
+ *   avoid the double free error.
+ * @tparam T represents the datatype to be used
+ * @tparam N represents the number of dimensions
+ */
+template <typename T, std::size_t N>
+MemRef<T, N> &MemRef<T, N>::operator=(MemRef<T, N> &&other) noexcept {
+  if (this != &other) {
+    // Free the original aligned and allocated space.
+    delete[] allocated;
+    // Steal members of the original object.
```

Comment:
Do you think you can overload this operator using the previous `Move` constructor itself?

Reply:
Used `move` constructor, concrete test cases are yet to be implemented.